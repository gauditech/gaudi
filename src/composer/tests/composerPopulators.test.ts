import { compile, compose, parse } from "@src/index";

describe("populator composer", () => {
  it("succeeds for simple populator", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }
      field description { type text }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        set is_new true // boolean literal setter
        set name "test name" // string literal setter
        set description hook { // hook setter
          arg name name // (translates to) changeset reference setter
          inline \`"Description of" + name\`
        }
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds for nested populators", () => {
    const bp = `
    model Org {
      field name { type text }

      relation repos { from Repo, through org }
    }

    model Repo {
      field name { type text }

      reference org { to Org }
      relation issues { from Issue, through repo }
    }

    model Issue {
      field title { type text }

      reference repo { to Repo }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        set name "test name"

        populate Repos {
          target relation repos as repo
  
          set name org.name // nested reference

          populate Issues {
            target relation issues as issue

            set title repo.name // nested reference
          }
        }
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds for fixed iteration", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        repeat 5

        set is_new true
        set name "test name"
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds for range iteration", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        repeat { start 1, end 3 }

        set is_new true
        set name "test name"
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds for range iteration (only end)", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        repeat { end 3 }

        set is_new true
        set name "test name"
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds for nested iterations", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type text }

      relation repos { from Repo, through org }
    }

    model Repo {
      field name { type text }

      reference org { to Org }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        repeat 4

        set is_new true
        set name "test name"

        populate repos {
          target relation repos as repo

          repeat { start 20, end 2000 }
          
          set name "test name"
        }
      }
    }`;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("succeeds with iterator variable in the context", () => {
    const bp = `
      model Org {
        field name { type text }
        field name2 { type text }
      }

      populator Dev {
        populate Orgs {
          repeat 10 as iter
          target model Org as org
          set name2 name
          set name hook {
            arg iter iter
            inline \`"Org " + iter.current\`
          }
        }
      }
      `;

    const def = compose(compile(parse(bp)));
    const populator = def.populators[0];

    expect(populator).toMatchSnapshot();
  });

  it("fails when missing field setter", () => {
    const bp = `
    model Org {
      field name { type text }
      field description { type text }
      field active { type boolean }
    }

    populator DevData {
      populate Orgs {
        target model Org as org

        set name "test name"
        // missing field setters for "description" and "active" fields
      }
    }`;

    expect(() => compose(compile(parse(bp)))).toThrowErrorMatchingInlineSnapshot(
      `"Action create-one "org" is missing setters for fields: description,active"`
    );
  });
});
