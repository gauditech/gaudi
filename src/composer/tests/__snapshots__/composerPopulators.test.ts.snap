// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`populator composer succeeds for fixed iteration 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "is_new",
              "setter": {
                "kind": "literal",
                "type": "boolean",
                "value": true,
              },
            },
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [],
      "repeater": {
        "alias": undefined,
        "end": 5,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds for nested iterations 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "is_new",
              "setter": {
                "kind": "literal",
                "type": "boolean",
                "value": true,
              },
            },
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [
            {
              "alias": "id",
              "kind": "field",
              "name": "id",
              "namePath": [
                "org",
                "id",
              ],
              "refKey": "Org.id",
            },
          ],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [
        {
          "actions": [
            {
              "alias": "repo",
              "changeset": [
                {
                  "name": "org_id",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "id",
                      ],
                      "alias": "org",
                    },
                    "type": "integer",
                  },
                },
                {
                  "name": "name",
                  "setter": {
                    "kind": "literal",
                    "type": "text",
                    "value": "test name",
                  },
                },
              ],
              "kind": "create-one",
              "model": "Repo",
              "select": [],
              "targetPath": [
                "repo",
              ],
            },
          ],
          "name": "repos",
          "populates": [],
          "repeater": {
            "alias": undefined,
            "end": 2000,
            "start": 20,
          },
          "target": {
            "alias": "repo",
            "identifyWith": {
              "name": "id",
              "paramName": "repo_id",
              "refKey": "Repo.id",
              "type": "integer",
            },
            "kind": "relation",
            "name": "repos",
            "namePath": [
              "Org",
              "repos",
            ],
            "refKey": "Org.repos",
            "retType": "Repo",
          },
        },
      ],
      "repeater": {
        "alias": undefined,
        "end": 4,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds for nested populators 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [
            {
              "alias": "id",
              "kind": "field",
              "name": "id",
              "namePath": [
                "org",
                "id",
              ],
              "refKey": "Org.id",
            },
            {
              "alias": "name",
              "kind": "field",
              "name": "name",
              "namePath": [
                "org",
                "name",
              ],
              "refKey": "Org.name",
            },
          ],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [
        {
          "actions": [
            {
              "alias": "repo",
              "changeset": [
                {
                  "name": "org_id",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "id",
                      ],
                      "alias": "org",
                    },
                    "type": "integer",
                  },
                },
                {
                  "name": "name",
                  "setter": {
                    "kind": "reference-value",
                    "target": {
                      "access": [
                        "name",
                      ],
                      "alias": "org",
                    },
                    "type": "text",
                  },
                },
              ],
              "kind": "create-one",
              "model": "Repo",
              "select": [
                {
                  "alias": "id",
                  "kind": "field",
                  "name": "id",
                  "namePath": [
                    "repo",
                    "id",
                  ],
                  "refKey": "Repo.id",
                },
                {
                  "alias": "name",
                  "kind": "field",
                  "name": "name",
                  "namePath": [
                    "repo",
                    "name",
                  ],
                  "refKey": "Repo.name",
                },
              ],
              "targetPath": [
                "repo",
              ],
            },
          ],
          "name": "Repos",
          "populates": [
            {
              "actions": [
                {
                  "alias": "issue",
                  "changeset": [
                    {
                      "name": "repo_id",
                      "setter": {
                        "kind": "reference-value",
                        "target": {
                          "access": [
                            "id",
                          ],
                          "alias": "repo",
                        },
                        "type": "integer",
                      },
                    },
                    {
                      "name": "title",
                      "setter": {
                        "kind": "reference-value",
                        "target": {
                          "access": [
                            "name",
                          ],
                          "alias": "repo",
                        },
                        "type": "text",
                      },
                    },
                  ],
                  "kind": "create-one",
                  "model": "Issue",
                  "select": [],
                  "targetPath": [
                    "issue",
                  ],
                },
              ],
              "name": "Issues",
              "populates": [],
              "repeater": {
                "end": 1,
                "start": 1,
              },
              "target": {
                "alias": "issue",
                "identifyWith": {
                  "name": "id",
                  "paramName": "issue_id",
                  "refKey": "Issue.id",
                  "type": "integer",
                },
                "kind": "relation",
                "name": "issues",
                "namePath": [
                  "Org",
                  "repos",
                  "issues",
                ],
                "refKey": "Repo.issues",
                "retType": "Issue",
              },
            },
          ],
          "repeater": {
            "end": 1,
            "start": 1,
          },
          "target": {
            "alias": "repo",
            "identifyWith": {
              "name": "id",
              "paramName": "repo_id",
              "refKey": "Repo.id",
              "type": "integer",
            },
            "kind": "relation",
            "name": "repos",
            "namePath": [
              "Org",
              "repos",
            ],
            "refKey": "Org.repos",
            "retType": "Repo",
          },
        },
      ],
      "repeater": {
        "end": 1,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds for range iteration (only end) 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "is_new",
              "setter": {
                "kind": "literal",
                "type": "boolean",
                "value": true,
              },
            },
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [],
      "repeater": {
        "alias": undefined,
        "end": 3,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds for range iteration 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "is_new",
              "setter": {
                "kind": "literal",
                "type": "boolean",
                "value": true,
              },
            },
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [],
      "repeater": {
        "alias": undefined,
        "end": 3,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds for simple populator 1`] = `
{
  "name": "DevData",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "is_new",
              "setter": {
                "kind": "literal",
                "type": "boolean",
                "value": true,
              },
            },
            {
              "name": "name",
              "setter": {
                "kind": "literal",
                "type": "text",
                "value": "test name",
              },
            },
            {
              "name": "description",
              "setter": {
                "args": [
                  {
                    "name": "name",
                    "setter": {
                      "kind": "changeset-reference",
                      "referenceName": "name",
                    },
                  },
                ],
                "code": {
                  "inline": ""Description of" + name",
                  "kind": "inline",
                },
                "kind": "fieldset-hook",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [],
      "repeater": {
        "end": 1,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;

exports[`populator composer succeeds with iterator variable in the context 1`] = `
{
  "name": "Dev",
  "populates": [
    {
      "actions": [
        {
          "alias": "org",
          "changeset": [
            {
              "name": "name",
              "setter": {
                "args": [
                  {
                    "name": "iter",
                    "setter": {
                      "kind": "context-reference",
                      "referenceName": "iter",
                    },
                  },
                ],
                "code": {
                  "inline": ""Org " + iter.current",
                  "kind": "inline",
                },
                "kind": "fieldset-hook",
              },
            },
            {
              "name": "name2",
              "setter": {
                "kind": "changeset-reference",
                "referenceName": "name",
              },
            },
          ],
          "kind": "create-one",
          "model": "Org",
          "select": [],
          "targetPath": [
            "org",
          ],
        },
      ],
      "name": "Orgs",
      "populates": [],
      "repeater": {
        "alias": "iter",
        "end": 10,
        "start": 1,
      },
      "target": {
        "alias": "org",
        "identifyWith": {
          "name": "id",
          "paramName": "org_id",
          "refKey": "Org.id",
          "type": "integer",
        },
        "kind": "model",
        "name": "Org",
        "namePath": [
          "Org",
        ],
        "refKey": "Org",
        "retType": "Org",
      },
    },
  ],
}
`;
